include/deal.II/lac/arpack_solver.h

```diff
@@ -491,49 +556,56 @@ void ArpackSolver::solve (const MatrixType1                  &/*system_matrix*/,
 
       int ldz = n;
 
-      std::vector<double> z (ldz*ncv, 0.);
-
       double sigmar = 0.0; // real part of the shift
       double sigmai = 0.0; // imaginary part of the shift
 
-      int lworkev = 3*ncv;
-      std::vector<double> workev (lworkev, 0.);
+      std::vector<double> eigenvalues_real (nev+1, 0.);
+      std::vector<double> eigenvalues_im (nev+1, 0.);
 
-      std::vector<double> eigenvalues_real (nev, 0.);
-      std::vector<double> eigenvalues_im (nev, 0.);
-
-      // call of ARPACK dneupd routine
-      dneupd_(&rvec, &howmany, &select[0], &eigenvalues_real[0],
-              &eigenvalues_im[0], &z[0], &ldz, &sigmar, &sigmai,
-              &workev[0], bmat, &n_inside_arpack, which, &nev, &tol,
-              &resid[0], &ncv, &v[0], &ldv,
-              &iparam[0], &ipntr[0], workd, &workl[0], &lworkl, &info);
+      // call of ARPACK dseupd/dneupd routine
+      if (additional_data.symmetric)
+        {
+          std::vector<double> z (ldz*nev, 0.);
+          dseupd_(&rvec, &howmany, &select[0], &eigenvalues_real[0],
+                  &z[0], &ldz, &sigmar, bmat, &n, which, &nev, &tol,
+                  &resid[0], &ncv, &v[0], &ldv,
+                  &iparam[0], &ipntr[0], &workd[0], &workl[0], &lworkl, &info);
+        }
+      else
+        {
+          std::vector<double> workev (3*ncv, 0.);
+          dneupd_(&rvec, &howmany, &select[0], &eigenvalues_real[0],
+                  &eigenvalues_im[0], &v[0], &ldz, &sigmar, &sigmai,
+                  &workev[0], bmat, &n, which, &nev, &tol,
+                  &resid[0], &ncv, &v[0], &ldv,
+                  &iparam[0], &ipntr[0], &workd[0], &workl[0], &lworkl, &info);
+        }
 
       if (info == 1)
         {
-          Assert (false, ExcArpackInfoMaxIt(control().max_steps()));
+          Assert (false, ArpackExcArpackInfoMaxIt(control().max_steps()));
         }
       else if (info == 3)
         {
-          Assert (false, ExcArpackNoShifts());
+          Assert (false, ArpackExcArpackNoShifts());
         }
       else if (info!=0)
         {
-          Assert (false, ExcArpackInfodneupd(info));
+          if (additional_data.symmetric)
+            {
+              Assert (false, ArpackExcArpackInfodseupd(info));
+            }
+          else
+            Assert (false, ArpackExcArpackInfodneupd(info));
         }
 
-
-      const unsigned int n_eigenvecs = eigenvectors.size();
-      for (size_type i=0; i<n_eigenvecs; ++i)
+      for (size_type i=0; i<nev; ++i)
         for (unsigned int j=0; j<n; ++j)
           eigenvectors[i](j) = v[i*n+j];
 
-      delete[] workd;
-
-      AssertDimension (eigenvalues.size(), eigenvalues_real.size());
-      AssertDimension (eigenvalues.size(), eigenvalues_im.size());
-
-      for (size_type i=0; i<eigenvalues.size(); ++i)
+      // nev might have been changed by plus one during dneupd
+      nev = (n_eigenvalues == 0) ? eigenvalues.size() : n_eigenvalues;
+      for (size_type i=0; i<nev; ++i)
```

Comment:
i guess this `for` was wrong initially in that `nev` is `int` anyway, so using `size_type` for counter is not consistent.